### PR TITLE
fix(liveness): Screen rotation saves/loads correctly; lifecycle not destroyed at time of use

### LINKED ISSUE #45: FaceLivenessDetector orientation
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui-android/issues?q=is%3Aissue+).

### Which UI component is this feature-request for?

Liveness

### Please include any relevant guides or documentation you're referencing

https://github.com/aws-amplify/amplify-ui-android/blob/main/liveness/src/main/java/com/amplifyframework/ui/liveness/ui/FaceLivenessDetector.kt

### Please describe your feature-request in detail.

The screen orientation lock function on line 102 of the FaceLivenessDetector.kt source could lock it in the app's current orientation rather than just in portrait mode. This way, the liveness screen could run in both orientations.

Currently if you run a LivenessSession in an application that runs in landscape orientation, the SDK forces the orientation to portrait mode causing the application to crash.

### Initialization steps (if applicable)

_No response_

### Code Snippet

```java
// Put your code below this line.

```

### amplifyconfiguration.json

_No response_

### Additional information and screenshots

image

### PR DESCRIPTION
**Pull request by tjleing**
Fix the crash in #45 by ensuring the lifecycle isn't destroyed.  Resolving that led to an issue where the screen would rotate infinitely, so we fix that by saving the intended restored orientation using `rememberSaveable`.  The existing `DisposableEffect` was being disposed on rotation, which restored the previous orientation immediately at that point.

Specifically, the crash was due to a race condition where:
1. rotation is landscape; `DisposableEffect` fires; requests Activity rotation
2. `LivenessCoordinator` is created which fires thread to bind camera to lifecycle
3. Activity rotation finishes
4. lifecycle is destroyed, crashing

- [x] PR conforms to [Pull Request](https://github.com/aws-amplify/amplify-ui-android/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guidelines.

*How did you test these changes?*
Tested locally on portrait and landscape apps.

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: liveness/src/main/java/com/amplifyframework/ui/liveness/camera/LivenessCoordinator.kt [hunk 1]
```diff
@@ -136,20 +137,22 @@ internal class LivenessCoordinator(
...
+                        bindToLifecycle(
+                            lifecycleOwner,
+                            CameraSelector.DEFAULT_FRONT_CAMERA,
+                            preview,
+                            analysis
+                        )
+                    } else {
+                        val faceLivenessException = FaceLivenessDetectionException(
+                            "A front facing camera is required but no front facing camera detected.",
+                            "Enable a front facing camera."
```
Comment (gpanshu) [inline liveness/src/main/java/com/amplifyframework/ui/liveness/camera/LivenessCoordinator.kt:152]: how would somone enable a front facing camera?

File: liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt [hunk 2]
```diff
@@ -18,19 +18,23 @@ package com.amplifyframework.ui.liveness.ui
...
+internal fun LockPortraitOrientation(content: @Composable (() -> Unit) -> Unit) {
     val context = LocalContext.current
-    DisposableEffect(Unit) {
-        val activity = context.findActivity() ?: return@DisposableEffect onDispose {}
-        val originalOrientation = activity.requestedOrientation
-        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
-        onDispose {
+    val activity = context.findActivity() ?: return content {}
+    val originalOrientation by rememberSaveable { mutableStateOf(activity.requestedOrientation) }
+    activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
```
Comment (mattcreaser) [inline liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt:33]: this is a side effect and should be in a `SideEffect` block or `LaunchedEffect(Unit)`

Comment (tylerjroach) [reply liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt:33]: I'm not fully following what the cause of the issue was in the first place. Maybe a key in the disposable would have worked?

File: liveness/src/main/java/com/amplifyframework/ui/liveness/camera/LivenessCoordinator.kt [hunk 3]
```diff
@@ -136,20 +137,22 @@ internal class LivenessCoordinator(
...
-                        preview,
-                        analysis
-                    )
-                } else {
-                    val faceLivenessException = FaceLivenessDetectionException(
-                        "A front facing camera is required but no front facing camera detected.",
-                        "Enable a front facing camera."
-                    )
-                    processSessionError(faceLivenessException, true)
+                if (lifecycleOwner.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
```
Comment (tjleing) [inline liveness/src/main/java/com/amplifyframework/ui/liveness/camera/LivenessCoordinator.kt:140]: check `!= .DESTROYED`, test out

File: liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt [hunk 4]
```diff
@@ -18,19 +18,23 @@ package com.amplifyframework.ui.liveness.ui
...
+
+    // wait until screen is rotated correctly
+    if (activity.requestedOrientation == ActivityInfo.SCREEN_ORIENTATION_PORTRAIT) {
```
Comment (tjleing) [inline liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt:36]: check actual rotation, not requested?

Comment (tjleing) [inline liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt:36]: test original with key as string instead of Unit

Comment (tjleing) [reply liveness/src/main/java/com/amplifyframework/ui/liveness/ui/LockPortraitOrientation.kt:36]: didn't work, sadly

**Review by gpanshu — APPROVED** [review]

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**